File: Arrays & String/Problem3 - 1431.java
import java.util.*;

public class Main {

    /* Brute Force Approach - (for every kid's find max) */
    public static List<Boolean> kidsWithCandies_BruteForce(int[] candies, int extraCandies) {
        int n = candies.length;
        List<Boolean> res = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            int ithKidTotalCandy = candies[i] + extraCandies;

            // find out max candies 
            int maxCandies = -1;
            for (int j = 0; j < n; j++) {
                maxCandies = Math.max(maxCandies, candies[j]);
            }

            res.add(ithKidTotalCandy >= maxCandies);
        }
        return res;
    }

    /* Optimal Approach - (externally count max element in the given array) */
    public static List<Boolean> kidsWithCandies_Optimal(int[] candies, int extraCandies) {
        int n = candies.length;
        List<Boolean> res = new ArrayList<>();

        // find out max candies 
        int maxCandies = -1;
        for (int i = 0; i < n; i++)
            maxCandies = Math.max(maxCandies, candies[i]);

        for (int i = 0; i < n; i++) {
            int ithKidTotalCandy = candies[i] + extraCandies;
            res.add(ithKidTotalCandy >= maxCandies);
        }

        return res;
    }

    public static void main(String[] args) {
        int[] candies = {2, 3, 5, 1, 3};
        int extraCandies = 3;

        List<Boolean> result1 = kidsWithCandies_BruteForce(candies, extraCandies);
        System.out.println("Brute Force Approach: " + result1);

        List<Boolean> result2 = kidsWithCandies_Optimal(candies, extraCandies);
        System.out.println("Optimal Approach: " + result2);
    }
}

File: Arrays & String/Problem3 - 1431.cpp
#include <bits/stdc++.h>
using namespace std;

/* Brute Force Approach - (for every kid's find max) */
vector<bool> kidsWithCandies(vector<int>& candies, int extraCandies) {
    int n = candies.size();
    vector<bool> res(n, 0);

    for (int i = 0; i < n; i++) {
        int ithKidTotalCandy = candies[i] + extraCandies;

        // find out max candies 
        int maxCandies = -1;
        for (int j = 0; j < n; j++) {
            maxCandies = max(maxCandies, candies[j]);
        }

        if (ithKidTotalCandy >= maxCandies) res[i] = 1;
    }
    return res;

    /* TC-O(n*n = n^2) SC-O(1) */
}

/* Optimal Approach - (externaly count max element in the give array) */
vector<bool> kidsWithCandies(vector<int>& candies, int extraCandies) {
    int n = candies.size();
    vector<bool> res(n, 0);

    // find out max candies 
    int maxCandies = -1;
    for (int i = 0; i < n; i++)
        maxCandies = max(maxCandies, candies[i]);

    for (int i = 0; i < n; i++) {
        int ithKidTotalCandy = candies[i] + extraCandies;
        if (ithKidTotalCandy >= maxCandies) {
            res[i] = 1;
        }
    }

    return res;
    /* TC-O(n + n = 2n) SC-O(1) */
}